AX: WebCore::endOfLine can return the same position it started with, causing an infinite loop when iterating over `AccessibilityObject::nextLineEndPosition`s

twilco opened the pull request:
#### 4b84c3ea8c264bb4451e08e0060d664212097c3b
<pre>
AX: WebCore::endOfLine can return the same position it started with, causing an infinite loop when iterating over `AccessibilityObject::nextLineEndPosition`s
<a href="https://bugs.webkit.org/show_bug.cgi?id=276656">https://bugs.webkit.org/show_bug.cgi?id=276656</a>
<a href="https://rdar.apple.com/131837344">rdar://131837344</a>

Reviewed by Andres Gonzalez.

In `WebCore::endOfLine(const VisiblePosition&amp; input)`, the input position is taken, and an `endPositionForLine` is
computed. It then checks to ensure that `inSameLine(input, computedEnd)`, and if not, it walks `previous()` one position
from `input` and gets the `endPositionForLine` of that position.

This all makes sense on the surface. `endOfLine`s job is to find the end position of the line `input` belongs to. The
problem is that `inSameLine` does not arguably work intuitively — it returns false if position A is editable, while
position B is not, even if they are rendered on the same line.

This mixed editability is present in markup on Google Docs documents:

&lt;span class=&quot;text-style&quot;&gt;Line of text&lt;/span&gt;&lt;span contenteditable=&quot;false&quot;&gt; &lt;br role=&quot;presentation&quot;/&gt;&lt;/span&gt;

There is a single space character before the br element, and if that position is compared with any earlier one,
`inSameLine` will return false.

I did try passing a flag through to `isSameLine` that ignored this editability requirement, but there are other ways
to trigger this behavior too, so the bug persisted. Instead, I&apos;ve made nextLineEndPosition resilient to this issue
by looping until we have moved off the start position.

Two testcases added — these hang indefinitely without the changes in this commit.

* LayoutTests/accessibility/mixed-contenteditable-double-br-visible-character-range-hang-expected.txt: Added.
* LayoutTests/accessibility/mixed-contenteditable-double-br-visible-character-range-hang.html: Added.
* LayoutTests/accessibility/mixed-contenteditable-visible-character-range-hang-expected.txt: Added.
* LayoutTests/accessibility/mixed-contenteditable-visible-character-range-hang.html: Added.
* LayoutTests/platform/glib/TestExpectations: Disable new tests.
* LayoutTests/platform/ios/TestExpectations: Enable new tests.
* LayoutTests/platform/ios/accessibility/mixed-contenteditable-double-br-visible-character-range-hang-expected.txt: Added.
* LayoutTests/platform/ios/accessibility/mixed-contenteditable-visible-character-range-hang-expected.txt: Added.
* LayoutTests/platform/mac-wk1/TestExpectations:
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::nextLineEndPosition const):

Canonical link: <a href="https://commits.webkit.org/281089@main">https://commits.webkit.org/281089@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4edc9ff5ae8cb0413003b35754bad3ae62b32e54

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/58636 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/37963 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/11121 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/62262 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/9080 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/60765 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/45600 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/9278 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/47452 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/6468 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/60667 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/35533 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/50695 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/28304 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/32278 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/7996 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/8084 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/54230 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/8268 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/63965 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/2549 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/8278 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/54773 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/2558 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/50721 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/54857 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12974 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/2140 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/33792 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/34878 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/35962 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/34623 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

AndresGonzalezApple commented on LayoutTests/accessibility/mixed-contenteditable-visible-character-range-hang.html:
```diff
@@ -0,0 +1,113 @@
+
+        // Ensure the main-thread isn't locked up by setting an attribute and checking the result.
+        document.getElementById("button").setAttribute("aria-label", "Foo bar");
+        await waitFor(() => platformTextAlternatives(accessibilityController.accessibleElementById("button")).includes("Foo bar"));
```
Is this necessary? I would think that if the main thread were locked, AXVisibleCharacterRange wouldn't return or would return the wrong value making the test fail.

twilco replied:
You're right, I think this was unnecessary. Removed in the latest commit.

AndresGonzalezApple commented on Source/WebCore/accessibility/AccessibilityObject.cpp:
```diff
@@ -1768,26 +1768,28 @@ String AccessibilityObject::stringForVisiblePositionRange(const VisiblePositionR
+    return lineEndPosition;
 }
 
 std::optional<VisiblePosition> AccessibilityObject::previousLineStartPositionInternal(const VisiblePosition& visiblePosition) const
```
Does this account for the case where you have two or more <br>s together?
If in the case of a line ending in a non-editable space followed by <br>, we return the <br> as the end of line position instead of the space, would the hang occur? I believe returning the <br> position is the right thing because what the client wants is the line range, and the end of that range should be past the last character.

twilco replied:
> If in the case of a line ending in a non-editable space followed by a br, we return the br as the end of line position instead of the space, would the hang occur?

It would, as the `<br>` is also wrapped by `contenteditable="false"`.

> Does this account for the case where you have two or more `br`s together?

I believe so, but I can add another testcase to be certain.

>  I believe returning the `br` position is the right thing because what the client wants is the line range, and the end of that range should be past the last character.

That makes sense. I think in general, this PR isn't changing the algorithm around what is the line end for any given position, just making sure we don't get stuck. I have tested this change on Google docs and VoiceOver output sounds good, so I think we should wait for a motivating example to investigate this bit further.

twilco committed:
AX: WebCore::endOfLine can return the same position it started with, causing an infinite loop when iterating over `AccessibilityObject::nextLineEndPosition`s
https://bugs.webkit.org/show_bug.cgi?id=276656
rdar://131837344

Reviewed by Andres Gonzalez.

In `WebCore::endOfLine(const VisiblePosition& input)`, the input position is taken, and an `endPositionForLine` is
computed. It then checks to ensure that `inSameLine(input, computedEnd)`, and if not, it walks `previous()` one position
from `input` and gets the `endPositionForLine` of that position.

This all makes sense on the surface. `endOfLine`s job is to find the end position of the line `input` belongs to. The
problem is that `inSameLine` does not arguably work intuitively — it returns false if position A is editable, while
position B is not, even if they are rendered on the same line.

This mixed editability is present in markup on Google Docs documents:

<span class="text-style">Line of text</span><span contenteditable="false"> <br role="presentation"/></span>

There is a single space character before the br element, and if that position is compared with any earlier one,
`inSameLine` will return false.

I did try passing a flag through to `isSameLine` that ignored this editability requirement, but there are other ways
to trigger this behavior too, so the bug persisted. Instead, I've made nextLineEndPosition resilient to this issue
by looping until we have moved off the start position.

Two testcases added — these hang indefinitely without the changes in this commit.

* LayoutTests/accessibility/mixed-contenteditable-double-br-visible-character-range-hang-expected.txt: Added.
* LayoutTests/accessibility/mixed-contenteditable-double-br-visible-character-range-hang.html: Added.
* LayoutTests/accessibility/mixed-contenteditable-visible-character-range-hang-expected.txt: Added.
* LayoutTests/accessibility/mixed-contenteditable-visible-character-range-hang.html: Added.
* LayoutTests/platform/glib/TestExpectations: Disable new tests.
* LayoutTests/platform/ios/TestExpectations: Enable new tests.
* LayoutTests/platform/ios/accessibility/mixed-contenteditable-double-br-visible-character-range-hang-expected.txt: Added.
* LayoutTests/platform/ios/accessibility/mixed-contenteditable-visible-character-range-hang-expected.txt: Added.
* LayoutTests/platform/mac-wk1/TestExpectations:
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::nextLineEndPosition const):

Canonical link: https://commits.webkit.org/281089@main
